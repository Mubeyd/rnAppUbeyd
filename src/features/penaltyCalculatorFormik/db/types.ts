export const AllowedBorrowingDays = 10;
export const PenaltyPerDay = 5;

export enum WeekendsType {
  Type1 = 1, // Saturday and Sunday
  Type2 = 2, // Friday and Saturday
}

export interface Contact {
  id: number;
  name: string;
}

export interface Country {
  id: number;
  name: string;
  currencySymbol: string;
  currencyName: string;
  flag: string;
  weekend: number;
  holiDays: string[];
  penaltyPerDay: number;
}

export interface IBookBorrow {
  bookBorrowDate: Date | null;
  bookReturnDate: Date | null;
  country: Country | null;
  bookPhotoFront: string | null;
  bookPhotoBack: string | null;
  randomText: string | undefined;
}
