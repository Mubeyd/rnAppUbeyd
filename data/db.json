{
  "contacts": [
    {
      "id": 1,
      "name": "John Doe"
    },
    {
      "id": 2,
      "name": "Mark Smith"
    },
    {
      "id": 3,
      "name": "Isabella Johnson"
    },
    {
      "id": 4,
      "name": "Fahrettin Demir"
    },
    {
      "id": 5,
      "name": "Mehmet Fatih"
    },
    {
      "id": 6,
      "name": "Pablo Hernandez"
    }
  ],
  "countries": [
    {
      "id": 1,
      "name": "Turkey",
      "currencySymbol": "₺",
      "currencyName": "Turkish Lira",
      "flag": "https://www.countryflags.io/tr/flat/64.png",
      "weekend": 1,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 5
    },
    {
      "id": 2,
      "name": "United Arab Emirates",
      "currencySymbol": "د.إ",
      "currencyName": "UAE Dirham",
      "flag": "https://www.countryflags.io/ae/flat/64.png",
      "weekend": 2,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 7
    },
    {
      "id": 3,
      "name": "Mexico",
      "currencySymbol": "$",
      "currencyName": "Mexican Peso",
      "flag": "https://www.countryflags.io/mx/flat/64.png",
      "weekend": 1,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 4
    },
    {
      "id": 4,
      "name": "United Kingdom",
      "currencySymbol": "£",
      "currencyName": "Pound Sterling",
      "flag": "https://www.countryflags.io/gb/flat/64.png",
      "weekend": 1,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 6
    },
    {
      "id": 5,
      "name": "Germany",
      "currencySymbol": "€",
      "currencyName": "Euro",
      "flag": "https://www.countryflags.io/de/flat/64.png",
      "weekend": 1,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 11
    },
    {
      "id": 6,
      "name": "United States of America",
      "currencySymbol": "$",
      "currencyName": "US Dollar",
      "flag": "https://www.countryflags.io/us/flat/64.png",
      "weekend": 1,
      "holiDays": ["01/01", "04/23", "05/01", "05/19", "07/12", "08/30", "10/29", "12/31"],
      "penaltyPerDay": 2
    }
  ]
}
