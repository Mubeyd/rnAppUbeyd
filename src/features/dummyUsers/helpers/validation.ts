import yup from 'yup';

export const validationSchema = yup.object().shape({
  firstName: yup.string().required('Please enter a first name'),
  lastName: yup.string().required('Please enter a last name'),
  age: yup.number().required('Please enter an age'),
  image: yup.string().required('Please enter an image'),
  companyName: yup.string().required('Please enter a company name'),
  companyTitle: yup.string().notRequired(),
  companyAddress: yup.string().notRequired(),
  companyCity: yup.string().notRequired(),
  companyState: yup.string().notRequired(),
  companyPostalCode: yup.string().notRequired(),
  companyDepartment: yup.string().notRequired(),
});
