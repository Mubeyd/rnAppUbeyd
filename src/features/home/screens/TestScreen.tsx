import { useEffect, useState } from 'react';
import { Button, Text, View } from 'react-native';

export default function TestScreen() {
  const [text, setText] = useState('');

  useEffect(() => {
    setTimeout(() => {
      // setText("timeout is called")
      //   console.log('timeout is called');
    }, 1000);
  }, []);

  return (
    <View>
      <Text>Test Screen</Text>
      <Text testID="myText">{text}</Text>
      <Button testID="myButton" title="Click me" onPress={() => setText('Hello world')} />
    </View>
  );
}
