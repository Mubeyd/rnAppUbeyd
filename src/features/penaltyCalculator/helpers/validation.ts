import { date, object, string } from 'yup';
import { Country, IBookBorrow } from '../db/types';

export const bookBorrowValidationSchema = object().shape({
  currentContact: object().shape({
    id: string().required('Please select a contact'),
    name: string().required('Please select a contact'),
  }),
  country: object<Country>().required('Please select a country'),
  bookBorrowDate: date().required('Please select a borrowing date'),
  bookReturnDate: date().required('Please select a returning date'),
  bookPhotoFront: string().required('Please take a photo of the front cover'),
  bookPhotoBack: string().required('Please take a photo of the back cover'),
});

// check validity

export const bookBorrowValidationSync = (bookBorrow: IBookBorrow) =>
  bookBorrowValidationSchema.isValidSync({ ...bookBorrow });

export const bookBorrowFieldsValidation = async (bookBorrow: IBookBorrow) => {
  try {
    await bookBorrowValidationSchema.validate(bookBorrow);
    return { isValid: true, errors: null };
  } catch (err: any) {
    return { isValid: false, errors: err };
  }
};

// you can try and type cast objects to the defined schema
export const bookBorrowFieldsCasting = (country: Country) =>
  bookBorrowValidationSchema.cast({
    ...country,
  });
