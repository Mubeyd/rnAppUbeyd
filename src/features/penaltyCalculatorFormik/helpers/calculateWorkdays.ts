import moment, { Moment } from 'moment';
import { WeekendsType } from '../db/types';

export function calculateWorkdays({
  startDate,
  endDate,
  weekendsType,
  holiDays,
}: {
  startDate: Moment;
  endDate: Moment;
  weekendsType: WeekendsType;
  holiDays: string[];
}) {
  const start = moment(startDate);
  const end = moment(endDate);
  let workdays = 0;

  for (let currentDay = start; currentDay.isSameOrBefore(end, 'day'); currentDay.add(1, 'day')) {
    if (weekendsType === WeekendsType.Type2) {
      if (isNotFridayANDSaturday(currentDay) && isNotHoliDay(currentDay, holiDays)) {
        workdays++;
      }
      continue;
    }
    if (isNotSaturdayANDSunday(currentDay) && isNotHoliDay(currentDay, holiDays)) {
      workdays++;
    }
  }

  return workdays;
}

const isNotSaturdayANDSunday = (currentDay: Moment) => {
  return currentDay.isoWeekday() !== 6 && currentDay.isoWeekday() !== 7;
};

const isNotFridayANDSaturday = (currentDay: Moment) => {
  return currentDay.isoWeekday() !== 5 && currentDay.isoWeekday() !== 6;
};

const isNotHoliDay = (currentDay: Moment, holiDays: string[]) => {
  return !holiDays.includes(currentDay.format('MM/DD'));
};
