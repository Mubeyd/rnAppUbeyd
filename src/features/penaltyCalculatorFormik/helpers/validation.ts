import { date, object, string } from 'yup';
// import { SchemaOf } from 'yup';
import { Contact, Country, IBookBorrow } from '../db/types';

export const bookBorrowValidationSchema = object<IBookBorrow>().shape({
  currentContact: object<Contact>().required('Please select a Contact'),
  country: object<Country>().required('Please select a country'),
  bookBorrowDate: date().required('Please select a borrowing date'),
  bookReturnDate: date().required('Please select a returning date'),
  bookPhotoFront: string().required('Please take a photo of the front cover'),
  bookPhotoBack: string().required('Please take a photo of the back cover'),
  randomText: string().required('Please enter some random text'),
});
