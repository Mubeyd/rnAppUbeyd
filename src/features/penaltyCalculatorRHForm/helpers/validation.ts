import { date, number, object, string } from 'yup';
import { Country, IBookBorrowForm } from '../db/types';

export const bookBorrowValidationSchema = object<IBookBorrowForm>().shape({
  country: object<Country>()
    .nullable()
    .shape({
      id: number().nullable().required('Please select a country'),
      name: string().nullable().required('Please select a country'),
    }),

  bookBorrowDate: date().nullable().required('Please select a borrowing date'),
  bookReturnDate: date().nullable().required('Please select a returning date'),
  bookPhotoFront: string().nullable().required('Please take a photo of the front cover'),
  bookPhotoBack: string().nullable().required('Please take a photo of the back cover'),
  randomText: string().min(1).nullable().required('Please enter some random text'),
});
