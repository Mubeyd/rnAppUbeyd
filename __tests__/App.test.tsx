/// UNIT TESTS
import moment from 'moment';
import { WeekendsType } from '../src/features/penaltyCalculatorFormik/db/types';
import { calculateWorkdays } from '../src/features/penaltyCalculatorFormik/helpers/calculateWorkdays';

it('calculates workdays', () => {
  expect(
    calculateWorkdays({
      startDate: moment('2021-01-01'),
      endDate: moment('2021-01-31'),
      weekendsType: WeekendsType.Type1,
      holiDays: ['01/01', '01/02', '01/03', '01/09', '01/10', '01/16', '01/17', '01/23', '01/24', '01/30', '01/31'],
    }),
  ).toBe(20);
});

/// UI TESTS
import axios from 'axios';
import React from 'react';
import { act, create } from 'react-test-renderer';
import TestScreen from '../src/features/home/screens/TestScreen';

const tree = create(<TestScreen />);

// jest.runAllTimers();

test('snapshot', () => {
  expect(tree).toMatchSnapshot();
});

// test('call timeout', () => {
//   act(() => jest.runAllTimers());
//   const text = tree.root.findByProps({ testID: 'myText' }).props;
//   expect(text.children).toEqual('timeout is called');
// });

test('button press', () => {
  const instance = tree.root;
  const button = instance.findByProps({ testID: 'myButton' }).props;
  act(() => button.onPress());

  const text = instance.findByProps({ testID: 'myText' }).props;
  expect(text.children).toEqual('Hello world');
});

/// Axios dummy users api test
test('request', async () => {
  const response = await axios.get('https://dummyjson.com/users');
  expect(response.status).toEqual(200);
  expect(response.data.users).toHaveLength(30);
});
