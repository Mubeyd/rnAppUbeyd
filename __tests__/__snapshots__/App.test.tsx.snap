// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<View>
  <Text>
    Test Screen
  </Text>
  <Text
    testID="myText"
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="myButton"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Click me
      </Text>
    </View>
  </View>
</View>
`;
